### ISSUE TITLE
Problem: some HTTP servers require Content-Length to be set to 0

### ISSUE THREAD

**Issue by yrashk** [omnigres/omnigres#279]
Currently, omni_httpd only sets `Content-Length` if the body is larger than zero bytes:

https://github.com/omnigres/omnigres/blob/master/extensions/omni_httpc/omni_httpc.c#L482

This is done similarly to H2O's HTTP client:

https://github.com/h2o/h2o/blob/master/src/httpclient.c#L484-L489

The HTTP/1.1 spec says:

<blockquote>
Applications SHOULD use this field to indicate the transfer-length of
   the message-body, unless this is prohibited by the rules in [section](https://www.rfc-editor.org/rfc/rfc2616#section-4.4)
   [4.4](https://www.rfc-editor.org/rfc/rfc2616#section-4.4).

...

   For compatibility with HTTP/1.0 applications, HTTP/1.1 requests
   containing a message-body MUST include a valid Content-Length header
   field unless the server is known to be HTTP/1.1 compliant. If a
   request contains a message-body and a Content-Length is not given,
   the server SHOULD respond with 400 (bad request) if it cannot
   determine the length of the message, or with 411 (length required) if
   it wishes to insist on receiving a valid Content-Length.

</blockquote>

Proposed solution 1: always include `Content-Length` (unless we're doing chunked transfers, of course)

Proposed solution 2: same as above, but provide an optional flag to suppress `Content-Length: 0` in the name of compaction

**Comment by steve-chavez** [comment]
> Currently, omni_httpd only sets Content-Length if the body is larger than zero bytes:

Hm, the above referred to `omni_httpc` and not `omni_httpd` right? I've tested `omni_httpd` and it always sets the `Content-Length` despite an empty response body. This is done here:

https://github.com/omnigres/omnigres/blob/e5ede5f0936ec1e02d765e557a405bfbc3a1fda3/extensions/omni_httpd/http_worker.c#L774

Would a test ensuring this behavior (`content-length` always set) be useful?

**Comment by yrashk** [comment]
You are right, the issue is about `omni_httpc`! That was my mistake, and thanks for spotting it!

**Comment by yrashk** [comment]
> Would a test ensuring this behavior (`content-length` always set) be useful?

On the httpc or the httpd side? Or both?

**Comment by steve-chavez** [comment]
>  httpd side?

Was thinking one for the `httpd` side first. Just to get me started :)

**Comment by yrashk** [comment]
I think there may be some interesting corner cases to test and codify.

Reviewing what I wrote suggests that a `null` body will lead to no `Content-Length`, I suppose?
https://github.com/omnigres/omnigres/blob/e5ede5f0936ec1e02d765e557a405bfbc3a1fda3/extensions/omni_httpd/http_worker.c#L758

There are some tests around content_length:

https://github.com/omnigres/omnigres/blob/e5ede5f0936ec1e02d765e557a405bfbc3a1fda3/extensions/omni_httpd/tests/http.yml#L132-L157

But they don't test the issue above or the error produced here:

https://github.com/omnigres/omnigres/blob/e5ede5f0936ec1e02d765e557a405bfbc3a1fda3/extensions/omni_httpd/http_worker.c#L765-L767

Since it's an error, it's mostly just trackable by seeing http 500, which is not perfect.

(In the near-term future, I wanted to have an extension like `omni_alerts` (or a better name?) that would allow to capture errors, custom alert messages from other components and notify the operator. When this will exist, testing against errors on the backend, in a background worker, may become nicer.)

Bottom line: improving the content-length related test coverage for omni_httpd is a good idea.

**Comment by steve-chavez** [comment]
> Reviewing what I wrote suggests that a null body will lead to no Content-Length, I suspect? That may not be true as my memory tells me h2o always sets it.

That turned out to be true as shown on https://github.com/omnigres/omnigres/pull/314/files#diff-d45d91bcc24a19bd5a9f92e23409dda685198f6195f9818ee88e66f9576ed554R195-R208.

> But they don't test the issue above or the error produced here:
> Since it's an error, it's mostly just trackable by seeing http 500, which is not perfect.

That produces a 200 OK + logs a WARNING as shown on https://github.com/omnigres/omnigres/pull/314/files#diff-d45d91bcc24a19bd5a9f92e23409dda685198f6195f9818ee88e66f9576ed554R167-R180, it doesn't fail with 500. Happy to change it to 500 if that's a better behavior - maybe it is, as the user might not notice the WARNING on the logs immediately.

**Comment by yrashk** [comment]
Let's continue the httpd side of the conversation in https://github.com/omnigres/omnigres/pull/314
